packages/transport/protocol-v2-e2e.ts

```diff
@@ -0,0 +1,538 @@
+import * as messages from '@trezor/protobuf/messages.json';
+import {
+    v2 as protocol,
+    thp as protocolThp,
+    TransportProtocolState,
+    TransportProtocol,
+} from '@trezor/protocol';
+import { createHash, randomBytes } from 'crypto';
+
+import { BridgeTransport, NodeUsbTransport, UdpTransport, Transport } from './src';
+
+function sha256(buffer: Buffer): Buffer {
+    return createHash('sha256').update(buffer).digest();
+}
+
+interface TransportState {
+    transport: Transport;
+    protocol: TransportProtocol;
+    path: any; // DeviceDescriptor
+    session: `${number}`;
+    channels: TransportProtocolState[];
+}
+
+interface Device {
+    transports: TransportState[];
+}
+
+const USE_BRIDGE = false;
+const USE_REAL_DEVICE = true;
+const PAIRING_METHODS = [1]; // [2, 3, 4];
+const USE_PAIRING: number = PAIRING_METHODS[0];
+
+/* eslint-disable @typescript-eslint/no-use-before-define */
+
+const run = async () => {
+    const ctrl = new AbortController();
+    let transport: Transport = USE_BRIDGE
+        ? new BridgeTransport({ messages, signal: ctrl.signal })
+        : new UdpTransport({ messages, signal: ctrl.signal });
+
+    if (USE_REAL_DEVICE) {
+        transport = new NodeUsbTransport({ messages, signal: ctrl.signal });
+    }
+
+    const transportState: TransportState = {
+        transport,
+        protocol,
+        path: '',
+        session: '0',
+        channels: [],
+    };
+
+    const device: Device = {
+        transports: [transportState],
+    };
+
+    let handshakeCommitment = Buffer.alloc(0);
+
+    const codeEntryPairing = async (code: number, state: any) => {
+        const hostKeys = protocolThp.getCpaceHostKeys(
+            code,
+            state.handshakeCredentials.handshakeHash,
+        );
+
+        const cpaceTrezor = await call('ThpCodeEntryCpaceHost', {
+            cpace_host_public_key: hostKeys.publicKey,
+        });
+
+        if (!cpaceTrezor.success) throw new Error('cpaceTrezor unsuccessful');
+
+        // 1. Set *shared_secret* = X25519(*cpace_host_private_key*, *cpace_trezor_public_key*).
+        // 2. Set *tag* = SHA-256(*shared_secret*).
+
+        const tag = protocolThp.getShareSecret(
+            Buffer.from(cpaceTrezor.payload.message.cpace_trezor_public_key, 'hex'),
+            hostKeys.privateKey,
+        );
+
+        const codeEntry = await call('ThpCodeEntryTag', {
+            tag,
+            // tag: Buffer.from(
+            //     '2278a1ba6b91e6d5faaecf0978340b55ab4c74824f0d45222cdc3d4a7395fc65',
+            //     'hex',
+            // ),
+        });
+
+        if (!codeEntry.success) {
+            throw new Error('ThpCodeEntryTag unsuccessful');
+        }
+
+        const { secret } = codeEntry.payload.message;
+        const sha = sha256(Buffer.from(secret, 'hex'));
+
+        console.warn('secret compare with commitment', sha.compare(handshakeCommitment));
+
```

Comment:
You should check that  *code* = SHA-256(*h* || *secret || challenge ||* PairingMethod_CodeEntry) % 1000000.

Reply:
Similarly for `nfcPairing` and `qrCodePairing`.